storage/gcloud/aio/storage/storage.py

```diff
@@ -129,22 +130,34 @@ def get_bucket(self, bucket_name: str) -> Bucket:
         log.debug('using %r gcloud storage upload method', upload_type)
 
         if upload_type == UploadType.SIMPLE:
-            return await self._upload_simple(url, object_name, data, headers,
+            return await self._upload_simple(url, object_name, stream, headers,
                                              session=session, timeout=timeout)
 
         if upload_type == UploadType.RESUMABLE:
-            return await self._upload_resumable(url, object_name, data,
+            return await self._upload_resumable(url, object_name, stream,
                                                 headers, session=session,
                                                 timeout=timeout)
 
         raise TypeError(f'upload type {upload_type} not supported')
 
     @staticmethod
-    def _preprocess_data(data: Any) -> Union[str, bytes]:
+    def _get_stream_len(stream: io.IOBase) -> int:
+        current = stream.tell()
+        try:
+            return stream.seek(0, os.SEEK_END)
+        finally:
+            stream.seek(current)
```

Comment:
Note:
Interesting use of the `try-except` block to rewind the stream. If you're curious, doing an assignment would be 2 instructions cheaper:
```
>>> def foo(data):
...     try:
...             return data
...     finally:
...             print(data)
... 
>>> dis.dis(foo)
  2           0 SETUP_FINALLY            4 (to 6)

  3           2 LOAD_FAST                0 (data)
              4 RETURN_VALUE

  5     >>    6 LOAD_GLOBAL              0 (print)
              8 LOAD_FAST                0 (data)
             10 CALL_FUNCTION            1
             12 POP_TOP
             14 END_FINALLY
             16 LOAD_CONST               0 (None)
             18 RETURN_VALUE
>>> def bar(data):
...     x = data
...     print(data)
...     return x
... 
>>> dis.dis(bar)
  2           0 LOAD_FAST                0 (data)
              2 STORE_FAST               1 (x)

  3           4 LOAD_GLOBAL              0 (print)
              6 LOAD_FAST                0 (data)
              8 CALL_FUNCTION            1
             10 POP_TOP

  4          12 LOAD_FAST                1 (x)
             14 RETURN_VALUE

```

Reply:
Good to know! In this case, I think its worth the extra ops, since that will additionally guarantee an exception raised by `stream.seek` does not affect whether we rewind or not.